### PR TITLE
add deepsearch-toolkit docling-ibm-models docling-parse docling-core mean-average-precision

### PR DESCRIPTION
**Pull request by hadim**
Checklist
- [x] Title of this PR is meaningful: e.g. "Adding my_nifty_package", not "updated meta.yaml".
- [x] License file is packaged (see [here](https://github.com/conda-forge/staged-recipes/blob/5eddbd7fc9d1502169089da06c3688d9759be978/recipes/example/meta.yaml#L64-L73) for an example).
- [x] Source is from official source.
- [x] Package does not vendor other packages. (If a package uses the source of another package, they should be separate packages or the licenses of all packages need to be packaged).
- [x] If static libraries are linked in, the license of the static library is packaged.
- [x] Package does not ship static libraries. If static libraries are needed, [follow CFEP-18](https://github.com/conda-forge/cfep/blob/main/cfep-18.md).
- [x] Build number is 0.
- [x] A tarball (`url`) rather than a repo (e.g. `git_url`) is used in your recipe (see [here](https://conda-forge.org/docs/maintainer/adding_pkgs.html#build-from-tarballs-not-repos) for more details).
- [x] GitHub users listed in the maintainer section have posted a comment confirming they are willing to be listed there.
- [x] When in trouble, please check our [knowledge base documentation](https://conda-forge.org/docs/maintainer/knowledge_base.html) before pinging a team.

### CODE REVIEW

**Comment by hadim** [comment]
Those are needed to package docling and deepsearch-glm but those two will come in different PRs as they are harder to package.

**Comment by hadim** [comment]
blocked by https://github.com/conda-forge/staged-recipes/pull/28092

**Comment by hadim** [comment]
I am seeing error related to missing `python_min` during the build. Is it because CEFP25 is not yet ready (https://github.com/conda-forge/conda-forge.github.io/pull/2355)?

I can see `python_min` defined at https://github.com/conda-forge/conda-forge-pinning-feedstock/blob/main/recipe/conda_build_config.yaml, but I don't know whether it is used in staged-recipes.

```
Adding in variants from /home/conda/staged-recipes-copy/.ci_support/linux64.yaml
Error:   × Failed to parse recipe

Error:   × Parsing: failed to parse match spec: Unable to parse version spec: .*
    ╭─[23:7]
 22 │   host:
 23 │     - python ${{ python_min }}.*
    ·       ─────────────┬────────────
    ·                    ╰── error parsing `python .*` as a match spec
 24 │     - poetry-core
    ╰────
Error:   × Parsing: failed to parse match spec: Unable to parse version spec: >=
    ╭─[27:7]
 26 │   run:
 27 │     - python >=${{ python_min }}
    ·       ─────────────┬────────────
    ·                    ╰── error parsing `python >=` as a match spec
 28 │     - urllib3 >=1.26.8
    ╰────
```

Sorry for the ping @beckermr but since you are the author of the various CEFP25 PRs, you might want to be aware of this.

**Comment by beckermr** [comment]
I am taking a look.

**Comment by hadim** [comment]
Note that the builds will fail anyway cuz of missing https://github.com/conda-forge/staged-recipes/pull/28092.

**Comment by beckermr** [comment]
the variable python_min is in the pinnings and rattler-build says it is loading it. I'd say unless we can reproduce with conda-build, this is a rattler bug.

**Comment by hadim** [comment]
I can reproduce this minimally indeed:

- `recipe.yaml`

```yaml
context:
  name: mean-average-precision
  version: 2024.01.05.0

package:
  name: ${{ name|lower }}
  version: ${{ version }}

build:
  script: echo ${{ python_min }}

requirements:
  host:
    - python ${{ python_min }}.*
  run:
    - python >=${{ python_min }}
```

- `config.yaml`

```yaml
python_min:
  # minimum supported python version per CFEP-25
  # bump to next minor version when we drop python versions
  - '3.9'
```

- Error:

```bash
❯ rattler-build build -m config.yaml -r recipe.yaml --render-only --output-dir /tmp/rattler_output
Error:   × Failed to parse recipe

Error:   × Parsing: failed to parse match spec: Unable to parse version spec: .*
    ╭─[14:7]
 13 │   host:
 14 │     - python ${{ python_min }}.*
    ·       ─────────────┬────────────
    ·                    ╰── error parsing `python .*` as a match spec
 15 │   run:
    ╰────
Error:   × Parsing: failed to parse match spec: Unable to parse version spec: >=
    ╭─[16:7]
 15 │   run:
 16 │     - python >=${{ python_min }}
    ·       ─────────────┬────────────
    ·                    ╰── error parsing `python >=` as a match spec
    ╰────
```


@wolfv is there is something obvious I am missing here or should I report this as a bug on the rattler-build repo?

**Comment by wolfv** [comment]
Yeah that's unfortunately a bug that we have to fix. I have made some preliminary work in https://github.com/prefix-dev/rattler-build/pull/1122 that better splits the two rendering phases:

- first find all unknown Jinja vars and render them out
- then render all expressions

**Comment by hadim** [comment]
> Yeah that's unfortunately a bug that we have to fix. I have made some preliminary work in [prefix-dev/rattler-build#1122](https://github.com/prefix-dev/rattler-build/pull/1122) that better splits the two rendering phases:
> 
> * first find all unknown Jinja vars and render them out
> * then render all expressions

Ok. Thanks for the input. I will hard-code it for now, waiting for rattler-build to support it.

**Comment by hadim** [comment]
@conda-forge/help-python @conda-forge/help-python-c ready for review.

- The failure on Windows is due to a missing pytorch.
- The lint error `python ={{ python_min }}` is a false positive, since with rattler-build we use `python ${{ python_min }}.*` instead of `python {{ python_min }}.*`.

---

@beckermr pinging you as you might want to be aware about the `python_min` linting false positive cuz of `rattler-build`.

**Comment by beckermr** [comment]
Thanks for the bump. I will fix.

**Comment by hadim** [comment]
@conda-forge/help-python @conda-forge/help-python-c, ready for review.

This is blocking for https://github.com/conda-forge/staged-recipes/pull/28093

The windows' failure is due to missing pytorch.

**Comment by hadim** [comment]
@conda-forge/help-python @conda-forge/help-python-c, let me know if you prefer to break down this PR into smaller ones.

**Comment by hadim** [comment]
@beckermr @wolfv can I ask one of you to be my hero here?

File: recipes/deepsearch-toolkit/recipe.yaml [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+  number: 0
+  noarch: python
+  script: python -m pip install . -vv --no-deps --no-build-isolation
+  python:
+    entry_points:
+      - deepsearch = deepsearch.cli:app
+
+requirements:
+  host:
+    - python ${{ python_min }}.*
```
Comment (beckermr) [inline recipes/deepsearch-toolkit/recipe.yaml:24]:
```suggestion
    - python ${{ python_min }}
```

File: recipes/docling-ibm-models/recipe.yaml [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+  sha256: 4ac155490a8646daca8aa7a45f0ee136ef96c4a0fbc24812694ec8fc45e5cadb
+
+build:
+  number: 0
+  noarch: python
+  script: python -m pip install . -vv --no-deps --no-build-isolation
+
+requirements:
+  host:
+    - python ${{ python_min }}.*
```
Comment (beckermr) [inline recipes/docling-ibm-models/recipe.yaml:21]:
```suggestion
    - python ${{ python_min }}
```

File: recipes/mean-average-precision/recipe.yaml [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+  sha256: b25dda0fdcf1f55d4ae8b9848fb1154369e87323bf472f9cd74da10cbaf992d0
+
+build:
+  number: 0
+  noarch: python
+  script: python -m pip install . -vv --no-deps --no-build-isolation
+
+requirements:
+  host:
+    - python ${{ python_min }}.*
```
Comment (beckermr) [inline recipes/mean-average-precision/recipe.yaml:21]:
```suggestion
    - python ${{ python_min }}
```

**Review by beckermr — APPROVED** [review]

File: recipes/deepsearch-toolkit/recipe.yaml (hunk 1 above)
Comment (hadim) [reply recipes/deepsearch-toolkit/recipe.yaml]: Apologize. I should have spotted that.

**Pull request merged**